src/bin/cargo/commands/add.rs

```diff
@@ -242,7 +242,24 @@ fn parse_dependencies(config: &Config, matches: &ArgMatches) -> CargoResult<Vec<
         .flatten()
         .map(|c| (Some(c.clone()), None))
         .collect::<IndexMap<_, _>>();
+
     let mut infer_crate_name = false;
+
+    for krate in crates.iter() {
+        let crate_name = krate.0.clone();
+        if let Some(crate_name) = crate_name {
```

Comment:
* Shorter.
* I doubt that we need to `clone` here.
* `unwrap` is safe. The `IndexMap` previously constructed has no `None` in the first tuple field.

```suggestion
    for (crate_name, _) in crates.iter() {
        let crate_name = crate_name.as_ref().unwrap();
```